### PR TITLE
feat: dprint formatter

### LINKED ISSUE #3818: replace prettier with dprint for "deno fmt"
https://dprint.dev/
https://github.com/dsherret/dprint/tree/master/packages/rust-dprint-plugin-typescript
dprint is a TS formatter written in Rust by David Sherret @dsherret

### PR DESCRIPTION
**Pull request by bartlomieju**
Closes #3818

### CODE REVIEW

**Comment by nayeemrmn** [comment]
Okay, the file matcher was designed like this:
- Accept a CSV list of globs in CLI. The default should be `["."]`.
- Expand these globs.
  - For files that are matched, add these files to the list of matches.
  - For directories that are matched, expand some list of "default" globs rooted on each directory. This time only match files. Add these files to the list of matches.
    - For the formatter, the default globs are `["**/*"]`.
    - For the test runner, the default globs are `["**/?(*_)test.{js,ts}"]`.
    - Both use just one glob pattern, so unless it's trivial you can support one default glob. But keep in mind that the patterns supported by just one glob pattern are quite limited.

See https://github.com/denoland/deno_std/pull/604 for the defence of this design.

**Comment by bartlomieju** [comment]
> Okay, the file matcher was designed like this:
> 
> * Accept a CSV list of globs in CLI. The default should be `["."]`.
> * Expand these globs.
>   
>   * For files that are matched, add these files to the list of matches.
>   * For directories that are matched, expand some list of "default" globs rooted on each directory. This time only match files. Add these files to the list of matches.
>     
>     * For the formatter, the default globs are `["**/*"]`.
>     * For the test runner, the default globs are `["**/?(*_)test.{js,ts}"]`.
>     * Both use just one glob pattern, so unless it's trivial you can support one default glob. But keep in mind that the patterns supported by just one glob pattern are quite limited.
> 
> See [denoland/deno_std#604](https://github.com/denoland/deno_std/pull/604) for the defence of this design.

Thanks for detailed explanation @nayeemrmn - PTAL at the PR, I added basic support for list of globs and formatter uses exact glob you specified above (though it should be limited to `.tsx?` and `.jsx?` as `dprint` does not support yet markdown). IIRC both formatter and test runner support ignore/exclude lists - I think I'll skip them in this PR and open an issue for that for later.

**Comment by nayeemrmn** [comment]
Edited my above comment with exclude semantics.

File: cli/formatter.rs [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+fn get_matching_files(glob_paths: Vec<String>) -> Vec<PathBuf> {
+  let mut target_files = Vec::with_capacity(128);
+
+  for path in glob_paths {
+    let files = glob::glob(&path)
+      .expect("Failed to execute glob.")
+      .filter_map(Result::ok);
+    target_files.extend(files);
+  }
```
Comment (nayeemrmn) [inline cli/formatter.rs]: Something like:
```rust
  for path in glob_paths {
    let matches = glob::glob(&path)
      .expect("Failed to execute glob.")
      .filter_map(Result::ok);
    for match in matches {
      if is_dir(&match) {
        let secondary_matches = glob::glob(&format("{}/**/*", &match))
          .expect("Failed to execute glob.")
          .filter_map(Result::ok);
        for secondary_match in secondary_matches {
          if !is_dir(&secondary_match) {
            target_files.push(secondary_match);
          }
        }
      } else {
        target_files.push(match);
      }
    }
  }
```

File: cli/tests/badly_formatted.js [hunk 2]
```diff
@@ -1,4 +1,4 @@
 
-console.log(
-  "Hello World"
+console.log("Hello World"
+
 )
```
Comment (ry) [inline cli/tests/badly_formatted.js]: Undo this change

File: cli/tests/integration_tests.rs [hunk 3]
```diff
@@ -29,12 +29,44 @@ fn fetch_test() {
...
+    .arg(badly_formatted_str)
+    .spawn()
+    .expect("Failed to spawn script")
+    .wait()
+    .expect("Failed to wait for child process");
+
+  assert_eq!(Some(0), status.code());
+  let expected = std::fs::read_to_string(fixed).unwrap();
+  let actual = std::fs::read_to_string(badly_formatted).unwrap();
+  assert_eq!(expected, actual);
```
Comment (ry) [inline cli/tests/integration_tests.rs]: Nice work

File: tools/fmt_test.py [hunk 4]
```diff
@@ -1,43 +0,0 @@
...
-                expected = f.read()
-            with open(dst) as f:
-                actual = f.read()
-            self.assertEqual(expected, actual)
-        finally:
-            shutil.rmtree(d)
-
-
-if __name__ == "__main__":
-    run_tests()
```
Comment (ry) [inline tools/fmt_test.py]: \o/

File: cli/tests/badly_formatted.js (hunk 2 above)
Comment (bartlomieju) [reply cli/tests/badly_formatted.js]: It's on purpose, `dprint` keeps string on separate line if you line break after parentheses...

File: cli/flags.rs [hunk 5]
```diff
@@ -556,155 +517,28 @@ The declaration file could be saved and used for typing information.",
 
 fn fmt_subcommand<'a, 'b>() -> App<'a, 'b> {
   SubCommand::with_name("fmt")
-        .about("Format files")
-        .long_about(
-"Auto-format JavaScript/TypeScript source code using Prettier
+    .about("Format files")
+    .long_about(
+      "Auto-format JavaScript/TypeScript source code using dprint
```
Comment (ry) [inline cli/flags.rs]: s/using dprint//
I don't think it's useful in the first sentence and probably confusing to people.

File: cli/formatter.rs [hunk 6]
```diff
@@ -0,0 +1,153 @@
+// Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
```
Comment (ry) [inline cli/formatter.rs]: Rename this file fmt.rs ?

Maybe give some //! explanation about how this works with links to dprint?

**Comment by bartlomieju** [comment]
@nayeemrmn @ry there are two flags that are not present in this PR that are pretty useful: `--stdout` and  `--ignore-path=<FILE...>`. Let's do them in follow up PR.

File: cli/tests/badly_formatted.js (hunk 2 above)
Comment (dsherret) [reply cli/tests/badly_formatted.js]: Opened https://github.com/dsherret/dprint/issues/61
